healthcheck/handler.go

```diff
@@ -0,0 +1,109 @@
+package healthcheck
+
+import (
+	"encoding/json"
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/pkg/errors"
+	"net/http"
+	"time"
+)
+
+const (
+	StatusOK       = "OK"
+	StatusCritical = "CRITICAL"
+	StatusWarning  = "WARNING"
+)
+
+// Handler returns the current health check of the app
+func (hc HealthCheck) Handler(w http.ResponseWriter, req *http.Request) {
+	now := time.Now().UTC()
+	ctx := req.Context()
+
+	var checks []Check
+
+	for _, client := range hc.Clients {
+		if client.Check != nil {
+			checks = append(checks, *client.Check)
+		}
+	}
+
+	hc.Status = hc.getStatus()
+	hc.Uptime = now.Sub(hc.StartTime)
+		hc.Checks =    checks
+
+
+	b, err := json.Marshal(hc)
+
+	if err != nil {
+		log.ErrorCtx(ctx, errors.Wrap(err, "failed to marshal json"), log.Data{"error": err, "health_check_response": hc})
+		return
+	}
+
+	_, err = w.Write(b)
+	if err != nil {
+		log.ErrorCtx(ctx, errors.Wrap(err, "failed to write bytes for http response"), nil)
+		return
+	}
+}
+
+// isAppStartingUp returns true or false based on if each client has had results back from at least a single check
+func (hc HealthCheck) isAppStartingUp() bool {
+	for _, client := range hc.Clients {
+		if client.Check == nil {
+			return true
+		}
+	}
+	return false
+}
+
+// getStatus returns a status as string as to the overall current apps health based on its dependent apps health
+func (hc HealthCheck) getStatus() string {
+	if hc.isAppStartingUp() {
+		return StatusWarning
+	}
+	return hc.isAppHealthy()
+}
+
+// isAppHealthy checks every clients check for their health then produces and returns a status for this apps health
+func (hc HealthCheck) isAppHealthy() string {
+	status := StatusOK
+	defer unlockAllClientChecks(hc.Clients)
+	for _, client := range hc.Clients {
```

Comment:
I wonder if it might be an idea to consider using go routines to _fan out_ here and execute the checks in parallel? Might be a performance optimisation but could be worth considering

Reply:
Happy to discuss further but I am thinking of leaving this, although I do like the thinking.
My two reasons are as follows: 

1. At the moment it is just reading a field from a struct and performing 1 conditional and 1 switch statement. So I believe this is a quick operation to complete already. It also returns if the Status on any of the `check`s is `StatusCritical`.

2. In order to do this the only part, we would change is the `appHealth := hc.readAppHealth(client)`  which would call off to `readAppHealth()` in its own goroutine and store the output. The issue is that goroutine can't really capture the value (technically possible but with numerous risks and issues) instead we would have to add channels to this section of code. This will make it more difficult to read and in this instance, I don't believe the optimisation payoff is worth the additional overhead of complexity.

        ```
var channels []chan string
	for _, client := range hc.Clients {
                chan := make(chan string)
                channels.append(channels, chan)
             
                status := StatusOK
	
		go hc.readAppHealth(client)
		var channels []chan string
	        var statuss []string
	        for _, client := range hc.Clients {
		        channel := make(chan string)
		        channels = append(channels, channel)
		        go hc.readAppHealth(client, channel)
		        status := StatusOK
	        }
	        for index, client := range hc.Clients {
		        select {
		        case statuss[index] := < channels[index]: //check if all statuss are set yet, if so call another function to then set this apps health 
		}
	}
	return status
```
Then in the readAppHealth that will have to pipe a message out to the channel once run and as per annotation a new function would ideally be made to cleaness (or function will be rather bloated).

As I say happy to discuss further on this